### Ana Luisa Espinoza López
**Expediente: 734753**

20/02/2023

# Tarea - Programación lineal

### Diseño de la Dieta Óptima

Se quiere producir comida para gatos de la manera más barata, no obstante se debe también asegurar que se cumplan los datos requeridos de analisis nutricional. Por lo que se quiere variar la cantidad de cada ingrediente para cumplir con los estandares nutricionales. Los requisitos que se tienen es que en $100$ gramos, se deben tener **por lo menos** $8$ gramos de proteína y $6$ gramos de grasa. Así mismo, no se debe tener más de $2$ gramos de fibra y $0.4$ gramos de sal.  

Los datos nutricionales se pueden obtener de la siguiente tabla:

Ingrediente|Proteína|Grasa|Fibra|Sal
:----|----|----|----|----
Pollo|  10.0%|08.0%|00.1%|00.2%
Carne|  20.0%|10.0%|00.5%|00.5%
Cordero|15.0%|11.0%|00.5%|00.7%
Arroz|  00.0%|01.0%|10.0%|00.2%
Trigo|  04.0%|01.0%|15.0%|00.8%
Gel|    00.0%|00.0%|00.0%|00.0%

Los costos de cada producto son:

Ingrediente|Costo por gramo
:----|----
Pollo|$\$$0.013
Carne|$\$$0.008
Cordero|$\$$0.010
Arroz|$\$$0.002
Trigo|$\$$0.005
Gel|$\$$0.001    

Lo que se busca optimizar en este caso es la cantidad de productos que se debe utilizar en la comida de gato, minimizando el costo total. Para simplificar la notación use las siguientes variables: 

+ $x_1:$ Gramos de pollo  
+ $x_2:$ Gramos de carne  
+ $x_3:$ Gramos de cordero  
+ $x_4:$ Gramos de arroz  
+ $x_5:$ Gramos de trigo  
+ $x_6:$ Gramos de gel

La tarea consiste en plantear el problemade programación lineal que permita satisfacer las necesidades alimenticias del gato al tiempo que minimice el costo total y resolverlo con `linprog`.

1. Criterio de optimización
    - Minimizar    
2. Variables
    + $x_i:$ gramos de cada ingrediente
3. Función de costo
    - $\min 0.013x_1+0.008x_2+0.01x_3+0.002x_4+0.005x_5+0.001x_6$    
4. Información adicional

    **Desigualdades**
    - $0.1x_1+0.2x_2+0.15x_3+0.04x_5 \geq 8$  
    (gramos de proteína por 100g de producto)
    - $0.08x_1+0.1x_2+0.11x_3+0.01x_4+0.01x_5 \geq 6$ 
    
    (gramos de grasa por 100g de producto)
    - $0.001x_1+0.005x_2+0.005x_3+0.1x_4+0.15x_5 \leq 2$ 
    
    (gramos de fibra por 100g de producto)
    - $0.002x_1+0.005x_2+0.007x_3+0.002x_4+0.008x_5 \leq 0.4$ 
    
    (gramos de sal por 100g de producto)
    
    **Igualdades**
    
    - $x_1+x_2+x_3+x_4+x_5+x_6=100$ (base de 100 gramos de producto)
 

In [2]:
import scipy.optimize as opt
import numpy as np

In [5]:
c=np.array([0.013,0.008,0.01,0.002,0.005,0.001])
A=np.array([
            [-0.1,-0.2,-0.15,0,-0.04,0],
            [-0.08,-0.1,-0.11,-0.01,-0.01,0],
            [0.001,0.005,0.005,0.1,0.15,0],
            [0.002,0.005,0.007,0.002,0.008,0]])

b=np.array([-8,-6,2,0.4])
Aeq=np.array([[1,1,1,1,1,1]])
beq=np.array([100])
opt.linprog(c,A,b,Aeq,beq)

     con: array([3.05107051e-11])
     fun: 0.5200000000216561
 message: 'Optimization terminated successfully.'
     nit: 8
   slack: array([4.00000000e+00, 7.70468134e-11, 1.70000000e+00, 1.00000000e-01])
  status: 0
 success: True
       x: array([4.50819667e-10, 6.00000000e+01, 4.08653405e-09, 2.29072691e-08,
       3.70730010e-10, 4.00000000e+01])

### Solución

Se minimizó el costo de la comida con un costo de \$0.520 por cada 100g de comida.

Para minimizar el costo de la comida de gato manteniendo los estándares nutricionales, se debe usar la siguiente cantidad de cada ingrediente:

    - pollo:   0g
    - carne:   60g
    - cordero: 0g
    - arroz:   0g
    - trigo:   0g
    - gel:     40g

### Problema de distribución de energía eléctrica

La Comisión Federal de Electricidad **(CFE)** dispone de cuatro plantas de generación para satisfacer la demanda diaria eléctrica en cuatro ciudades, Guadalajara, León y Morelia. Las plantas $1$, $2$ y $3$ pueden satisfacer $80$, $40$ y $60$ millones de KW al día respectivamente. Las necesidades de las ciudades de Guadalajara, León y Morelia son de $70$, $40$ y $70$ millones de Kw al día respectivamente. 


Los costos asociados al envío de suministro energético por cada millón de Kw entre cada planta y cada ciudad son los registrados en la siguiente tabla. 

-|Guadalajara|León|Morelia
:----|----|----|----
Planta 1|5|2|7
Planta 2|3|6|6
Planta 3|6|1|2

Y por último, las restricciones del problema, van a estar dadas por las capacidades de oferta y demanda de cada planta (en millones de KW) y cada ciudad.

Para simplificar la notación use las siguientes variables:

+ $x_1$: Kw (en millones) distribuidos de la Planta 1 a Guadalajara
+ $x_2$: Kw (en millones) distribuidos de la Planta 1 a León
+ $x_3$: Kw (en millones) distribuidos de la Planta 1 a Morelia
+ $x_4$: Kw (en millones) distribuidos de la Planta 2 a Guadalajara
+ $x_5$: Kw (en millones) distribuidos de la Planta 2 a León
+ $x_6$: Kw (en millones) distribuidos de la Planta 2 a Morelia
+ $x_7$: Kw (en millones) distribuidos de la Planta 3 a Guadalajara
+ $x_8$: Kw (en millones) distribuidos de la Planta 3 a León
+ $x_9$: Kw (en millones) distribuidos de la Planta 3 a Morelia

La tarea consiste en plantear el problema de programación lineal que permita satisfacer las necesidades de todas las ciudades al tiempo que minimice los costos asociados a la distribución y resolverlo con `linprog`.

1. Criterio de optimización
    - Minimizar    
2. Variables
    - Kw en millones distribuidos de cada planta a cada ciudad $x_i$
3. Función de costo
    - $5x_1+2x_2+7x_3+3x_4+6x_5+6x_6+6x_7+x_8+2x_9$    
4. Información adicional
    + $x_1+x_2+x_3 \leq 80$ (Capacidad planta 1)
    + $x_4+x_5+x_6 \leq 40$ (Capacidad planta 2)
    + $x_7+x_8+x_9 \leq 60$ (Capacidad planta 3)
    + $x_1+x_4+x_7 \geq 70$ (Demanda Guadalajara)
    + $x_2+x_5+x_8 \geq 40$ (Demanda León)
    + $x_3+x_6+x_9 \geq 70$ (Demanda Morelia)

In [7]:
c=np.array([5,2,7,3,6,6,6,1,2])
A=np.array([[1,1,1,0,0,0,0,0,0],
            [0,0,0,1,1,1,0,0,0],
            [0,0,0,0,0,0,1,1,1],
            [-1,0,0,-1,0,0,-1,0,0],
            [0,-1,0,0,-1,0,0,-1,0],
            [0,0,-1,0,0,-1,0,0,-1]])
b=np.array([80,40,60,-70,-40,-70])


In [8]:
opt.linprog(c,A,b)

     con: array([], dtype=float64)
     fun: 539.9999999945555
 message: 'Optimization terminated successfully.'
     nit: 6
   slack: array([ 8.82408813e-10,  4.27498037e-10,  6.52505605e-10, -7.68807240e-10,
       -4.24719815e-10, -7.68878294e-10])
  status: 0
 success: True
       x: array([3.00000000e+01, 4.00000000e+01, 1.00000000e+01, 4.00000000e+01,
       4.66958755e-11, 6.07352053e-12, 2.20740129e-11, 2.37487168e-11,
       6.00000000e+01])

### Solución
Para minimizar el costo de envió por \$540 por cada millón de Kw enviado, cada planta debe enviar lo siguiente:

**La planta 1 distribuirá:**
- 30 millones de Kw a Guadalajara
- 40 millones de Kw a León
- 10 millones de Kw a Morelia

**La planta 2 distribuirá:**
- 40 millones de Kw a Guadalajara
- 0 millones de Kw a León
- 0 millones de Kw a Morelia

**La planta 3 distribuirá:**
- 0 millones de Kw a Guadalajara
- 0 millones de Kw a León
- 60 millones de Kw a Morelia